katdal/test/test_chunkstore_s3.py

```diff
@@ -175,34 +175,75 @@ def test_jwt_broken_token(self):
         h, p, s = token.split('.')
-        assert_raises(InvalidToken, decode_jwt, '.'.join((h, p[:-1], s)))
+        with pytest.raises(InvalidToken, match=r'Could not decode'):
+            decode_jwt('.'.join((h, p[:-1], s)))
         # Token signature failed to decode or wrong length
-        assert_raises(InvalidToken, decode_jwt, token[:-1])
-        assert_raises(InvalidToken, decode_jwt, token[:-2])
-        assert_raises(InvalidToken, decode_jwt, token + token[-4:])
+        with pytest.raises(InvalidToken, match=r'too short'):
+            decode_jwt(token[:-1])
+        with pytest.raises(InvalidToken, match=r'too short'):
+            decode_jwt(token[:-2])
+        with pytest.raises(InvalidToken, match=r'too long'):
+            decode_jwt(token + token[-4:])
 
     def test_jwt_expired_token(self):
         payload = {'exp': 0, 'iss': 'kat', 'prefix': ['123']}
-        token = encode_jwt(payload)
-        assert_raises(InvalidToken, decode_jwt, token)
+        with pytest.raises(InvalidToken, match=r'Token expired at 01-Jan-1970'):
+            decode_jwt(encode_jwt(payload))
         # Check that expiration time is not-too-large integer
-        payload['exp'] = 1.2
-        assert_raises(InvalidToken, decode_jwt, encode_jwt(payload))
+        payload['exp'] = '1.2'
+        with pytest.raises(InvalidToken, match=r'Expiration time must be an integer'):
+            decode_jwt(encode_jwt(payload))
         payload['exp'] = 12345678901234567890
-        assert_raises(InvalidToken, decode_jwt, encode_jwt(payload))
+        with pytest.raises(InvalidToken, match=r'an integer that is not too large'):
+            decode_jwt(encode_jwt(payload))
         # Check that it works without expiry date too
         del payload['exp']
         claims = decode_jwt(encode_jwt(payload))
-        assert_equal(payload, claims)
+        assert payload == claims
+
+
+class UnexpectedDuration(AssertionError):
+    """Test took either too short or too long to complete."""
+
+
+def duration(minimum, maximum=None):
+    """Test must finish within `minimum` to `maximum` seconds to pass.
+
+    Example use::
+
+      @duration(.2, .3)
+      def test_that_fails():
+          time.sleep(.4)
+
+    The default `maximum` differs from `minimum` by a small tolerance.
+    The check only happens if the module-level variable `check_durations` is True.
+    """
+    if maximum is None:
+        maximum = minimum + TEST_DURATION_TOLERANCE
+
+    def decorate(func):
+        def newfunc(*arg, **kw):
+            start = time.time()
+            result = func(*arg, **kw)
+            duration = time.time() - start
+            if check_durations and not minimum <= duration <= maximum:
+                raise UnexpectedDuration(f"Test took {duration:g} seconds, which is "
+                                         f"outside the range [{minimum:g}, {maximum:g}]")
+            return result
+        return newfunc
+    return decorate
```

Comment:
This is probably fine (modulo Richard's comment about functools.wraps), but if you wanted a pytest-onic way to do this it would be to use a `pytest.mark` to specify the range for each test, and one of the pytest hook functions to examine the mark and override how the test is run. As far as I recall the hook functions have access to the config object, so can examine the configuration without needing a nasty global variable.

Reply:
So would this work with a method too (looks like it)? That was why I ended up with old-school decorators instead of fixtures.

Reply:
I would imagine so - marks can also be set on a class to affect all the methods in the class. I've not actually looked into the hooks that affect running a test, so I'm a little hazy on what the implementation would actually look like.

There is a package called pytest-timeout that might be interesting to look at as an example. It applies just an upper limit, although has a lot more complexity so that it can kill the test if it exceeds the limit.

Reply:
Thanks Bruce, you've now got me hooked on hooks 😛  and I have a nifty solution in the wings.

I did look at pytest-timeout but it's a different usecase. It basically wants to terminate tests that take too long, with all its concurrency challenges. I want to know that the test takes "exactly" as long as expected, to test my assumptions of the retry mechanisms.